### PR TITLE
Rewrite Client-side: Backbone -> Vue

### LINKED ISSUE #543: Bookmark description as card link tooltip
_[Just opening a discussion with few thoughts on bookmark descriptions accessibility]_

Bookmarks descriptions can sometimes be quite important when we add personal notes to a bookmark to recontextualise it. Currently, viewing the description requires clicking on the bookmark (which is now already way more fluid than before).

What about further improving this by also displaying bookmark's description when hovering the bookmark link ? :)  (yes, this idea directly conflicts with current tooltips simple indications)

It could be done with native tooltip, line 3 in `bookmarks/js/templates/BookmarkCard.html`, or with a more beautiful design that could be kind of inspired from https://jqueryui.com/tooltip/ ?

When hovering on the bookmark card link, the `'Open website in new tab'` tooltip could be transformed into something like: `'This is the bookmark description blablabla.... *LINE BREAK*
Click here to open website in a new tab'`

Any thoughts on this @marcelklehr ? :)

### LINKED ISSUE #633: Translate bot forgets a string
"There are no bookmarks available for this query. Try changing your filter or add some using the menu entry on the left."
The sentence is currently not translated and shown in english when using the german interface.

see https://github.com/nextcloud/docker-ci/issues/125

### LINKED ISSUE #694: Translations missing?
https://help.nextcloud.com/t/strings-in-english-source-not-complete/52374

### LINKED ISSUE #747: Menu items on iOS browsers are not working
### Steps to reproduce
1. Sync Chrome bookmarks with Nextcloud using Floccus
2. Open Nexcloud Bookmarks with Safari or Chrome on iOS (iPad Air and iPhone X)
3. Try to use folders
3.1 Click on a folder to show bookmarks
3.2 Click the arrow on the left to show subfolders
3.3 Click the three dots on the right to show options

### Expected behaviour
1. Bookmarks in a selected folder should be listed
2. Subfolders should be listed under the selected folder
3. Folder options should show up

### Actual behaviour
1. No bookmarks shown, but message "No bookmarks here."
2. Subfolders are not showing up when clicking on the arrow. It looks like it's just enabling drag&drop instead.
3. On the iPhone the dots are hidden and are just visible when moving the page, but then you cannot click on them. On the iPad the dots are shown but cannot be clicked.

On desktop browsers it's working as expected.

### Server configuration

**Operating system**: Linux, Docker

**Web server:** Apache

**Database:** MySQL

**Nextcloud version:** 16.0.3

**Bookmarks version:** 1.0.6

### Client configuration
**Browser:** Safari and Chrome 75.0.3770.103

**Operating system:** iOS 12.3.1

### LINKED ISSUE #756: Cannot tell from webui if folder is empty or contains sub-folders unless moused over
Hi!  I cannot tell if a folder contains subfolders,

unless I directly mouse over it.

This becomes confusing if the selected folder contains no bookmarks, but the sub-folder(s) contain bookmarks.  This message will be displayed:

This is especially confusing to users attempting to delete folders they assume are empty.

### LINKED ISSUE #772: Moving bookmark from root to folder keeps bookmark in root folder
### Steps to reproduce
1. Create a new bookmark for http://example.com/
2. Create a new folder 'Test'
3. Drag the bookmark from step 1 into folder Test

### Expected behaviour
The bookmark should be in the Test folder, and only in the Test folder.

### Actual behaviour
The bookmark is shown in the Test folder, but is also still in the 'root' folder, as seen from the database:
```
MariaDB [nextcloud]> select * from oc_bookmarks where url='http://example.com/';
+----+---------------------+----------------+---------+-------------+--------+------------+--------------+------------+--------------+
| id | url                 | title          | user_id | description | public | added      | lastmodified | clickcount | last_preview |
+----+---------------------+----------------+---------+-------------+--------+------------+--------------+------------+--------------+
| 49 | http://example.com/ | Example Domain | gmc     |             |      0 | 1565944145 |   1565944153 |          0 |            0 |
+----+---------------------+----------------+---------+-------------+--------+------------+--------------+------------+--------------+
1 row in set (0.01 sec)

MariaDB [nextcloud]> select * from oc_bookmarks_folders_bookmarks where bookmark_id=49;
+-----------+-------------+-------+
| folder_id | bookmark_id | index |
+-----------+-------------+-------+
|        -1 |          49 |    41 |
|         7 |          49 |     2 |
+-----------+-------------+-------+
2 rows in set (0.00 sec)

MariaDB [nextcloud]> 
```

## Server configuration detail

**Operating system:** Linux 4.14.133-166 #1 SMP PREEMPT Mon Aug 5 22:49:45 -03 2019 armv7l

**Webserver:** nginx/1.17.2 (fpm-fcgi)

**Database:** mysql 10.0.38

**PHP version:** 

7.3.8
Modules loaded: Core, date, libxml, openssl, pcre, sqlite3, zlib, ctype, curl, dom, fileinfo, filter, ftp, hash, iconv, json, mbstring, SPL, PDO, session, posix, readline, Reflection, standard, SimpleXML, pdo_sqlite, Phar, tokenizer, xml, xmlreader, xmlwriter, mysqlnd, cgi-fcgi, apcu, exif, gd, imagick, intl, ldap, memcached, pcntl, pdo_mysql, pdo_pgsql, redis, sodium, zip, Zend OPcache

**Nextcloud version:** 16.0.4 - 16.0.4.1

**Updated from an older Nextcloud/ownCloud or fresh install:** 

**Where did you install Nextcloud from:** unknown

<details><summary>Signing status</summary>

Array
(
)

</details>

<details><summary>List of activated apps</summary>

```
Enabled:
 - accessibility: 1.2.0
 - activity: 2.9.1
 - admin_audit: 1.6.0
 - bookmarks: 1.1.1
 - bruteforcesettings: 1.4.0
 - calendar: 1.7.0
 - cloud_federation_api: 0.2.0
 - comments: 1.6.0
 - contacts: 3.1.3
 - dav: 1.9.2
 - federatedfilesharing: 1.6.0
 - federation: 1.6.0
 - files: 1.11.0
 - files_external: 1.7.0
 - files_pdfviewer: 1.5.0
 - files_rightclick: 0.13.0
 - files_sharing: 1.8.0
 - files_texteditor: 2.8.0
 - files_trashbin: 1.6.0
 - files_versions: 1.9.0
 - files_videoplayer: 1.5.0
 - firstrunwizard: 2.5.0
 - gallery: 18.3.0
 - issuetemplate: 0.5.0
 - logreader: 2.1.0
 - lookup_server_connector: 1.4.0
 - news: 13.1.6
 - nextcloud_announcements: 1.5.0
 - notes: 3.0.1
 - notifications: 2.4.1
 - oauth2: 1.4.2
 - password_policy: 1.6.0
 - previewgenerator: 2.1.0
 - privacy: 1.0.0
 - provisioning_api: 1.6.0
 - recommendations: 0.4.0
 - serverinfo: 1.6.0
 - sharebymail: 1.6.0
 - support: 1.0.0
 - survey_client: 1.4.0
 - systemtags: 1.6.0
 - tasks: 0.11.1
 - text: 1.0.2
 - theming: 1.7.0
 - twofactor_backupcodes: 1.5.0
 - updatenotification: 1.6.0
 - viewer: 1.1.0
 - workflowengine: 1.6.0
Disabled:
 - encryption
 - user_ldap

```
</details>

<details><summary>Configuration (config/config.php)</summary>

```
{
    "memcache.local": "\\OC\\Memcache\\APCu",
    "apps_paths": [
        {
            "path": "\/var\/www\/html\/apps",
            "url": "\/apps",
            "writable": false
        },
        {
            "path": "\/var\/www\/html\/custom_apps",
            "url": "\/custom_apps",
            "writable": true
        }
    ],
    "instanceid": "***REMOVED SENSITIVE VALUE***",
    "passwordsalt": "***REMOVED SENSITIVE VALUE***",
    "secret": "***REMOVED SENSITIVE VALUE***",
    "trusted_domains": [
        "example.com",
    ],
    "datadirectory": "***REMOVED SENSITIVE VALUE***",
    "overwrite.cli.url": "https:\/\/www.example.com",
    "dbtype": "mysql",
    "version": "16.0.4.1",
    "dbname": "***REMOVED SENSITIVE VALUE***",
    "dbhost": "***REMOVED SENSITIVE VALUE***",
    "dbport": "",
    "dbtableprefix": "oc_",
    "mysql.utf8mb4": true,
    "dbuser": "***REMOVED SENSITIVE VALUE***",
    "dbpassword": "***REMOVED SENSITIVE VALUE***",
    "installed": true,
    "maintenance": false,
    "loglevel": 2,
    "app_install_overwrite": [
        "calendar"
    ]
}
```
</details>

**Are you using external storage, if yes which one:** local

**Are you using encryption:** 

**Are you using an external user-backend, if yes which one:** no

## Client configuration

**Browser:** Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:68.0) Gecko/20100101 Firefox/68.0

**Operating system:** 

## Logs

<details><summary>Web server error log</summary>

```
n/a
```
</details>

<details><summary>Nextcloud log</summary>

```
n/a
```
</details>

<details><summary>Browser log</summary>

n/a
</details>

### LINKED ISSUE #784: Wierd problems with limits of subfolders with longer names.
Screenshot from 2019-08-29 11-14-38

_In the picture depicted above, the "Server Management" sub-folder was shortened to "Server Mgmt" in order to stop some problems I describe below._

### Steps to reproduce 
1. Make multiple sub-folder levels with long names.

### Expected behavior
The functionality of shrink/expand, select, rename, etc should work normally on all folder objects.

### Actual behavior
As the folder's visible object gets further to the right where the left side-panel ends, the functionality gets wonky. Where you're supposed to click to expand/collapse changes from the intended location to further right where select should be. The ability to select a folder with subfolders can go away. The ability to rename a folder/subfolder seems there but to gain focus on the textbox for renaming is inaccessible.

### Possible solutions
1. Make the border separating the left side-panel moveable.
2. Add a slider to the bottom of the left side-panel for when the contents render past the edge.
3. Display an abbreviated folder name unless it's selected.
4. Make the font size of a folder name dependent on the character count.
5. Have the long folder names break over multiple lines making the folder selection object taller.
6. Tell users to stop making long folder names and to stop having so much nesting.

### Server configuration
The `apps/issuetemplate` isn't a thing anymore, and my button for "Generate System Report" in `index.php/settings/admin/support` isn't working.

**Operating system:** Ubuntu Linux 18.04.1

**Web server:** Apache 2.4.29

**Database:** MySQL 5.7.27

**PHP version:** 7.2.19

**Nextcloud version:** (see Nextcloud admin page) 16.0.4

**Bookmarks version:** (see Nextcloud apps page) 1.1.1

**Updated from an older Nextcloud or fresh install:** Updated.

**Signing status:** ?

```
No errors have been found.
```

**List of activated apps:**

```
Enabled:
  - accessibility: 1.2.0
  - activity: 2.9.1
  - admin_audit: 1.6.0
  - announcementcenter: 3.5.1
  - apporder: 0.7.1
  - audioplayer: 2.7.2
  - bookmarks: 1.1.1
  - bruteforcesettings: 1.4.0
  - calendar: 1.7.0
  - camerarawpreviews: 0.7.0
  - carnet: 0.17.0
  - checksum: 0.4.3
  - circles: 0.17.7
  - cloud_federation_api: 0.2.0
  - comments: 1.6.0
  - contacts: 3.1.3
  - cookbook: 0.3.2
  - cospend: 0.1.0
  - data_request: 1.3.0
  - dav: 1.9.2
  - dicomviewer: 1.2.0
  - drawio: 0.9.3
  - drop_account: 0.1.0
  - event_update_notification: 0.3.4
  - extract: 1.2.0
  - federatedfilesharing: 1.6.0
  - federation: 1.6.0
  - files: 1.11.0
  - files_accesscontrol: 1.6.0
  - files_downloadactivity: 1.5.0
  - files_ebookreader: 0.0.1
  - files_markdown: 2.0.6
  - files_mindmap: 0.0.15
  - files_pdfviewer: 1.5.0
  - files_photospheres: 1.0.5
  - files_readmemd: 1.1.1
  - files_retention: 1.5.1
  - files_rightclick: 0.15.1
  - files_sharing: 1.8.0
  - files_trashbin: 1.6.0
  - files_versions: 1.9.0
  - files_videoplayer: 1.5.0
  - firstrunwizard: 2.5.0
  - flowupload: 0.1.2
  - forms: 1.0.3
  - gallery: 18.3.0
  - gpxedit: 0.0.11
  - gpxmotion: 0.0.9
  - gpxpod: 4.0.5
  - groupfolders: 4.1.0
  - guests: 1.1.1
  - impersonate: 1.3.0
  - logreader: 2.1.0
  - lookup_server_connector: 1.4.0
  - mail: 0.16.0
  - metadata: 0.9.0
  - music: 0.10.0
  - nextcloud_announcements: 1.5.0
  - notes: 3.0.1
  - notifications: 2.4.1
  - oauth2: 1.4.2
  - ocsms: 2.1.3
  - password_policy: 1.6.0
  - passwords: 2019.8.1
  - phonetrack: 0.5.2
  - polls: 0.10.2
  - previewgenerator: 2.1.0
  - privacy: 1.0.0
  - provisioning_api: 1.6.0
  - quota_warning: 1.5.0
  - radio: 0.6.5
  - rainloop: 6.0.3
  - ransomware_detection: 0.5.2
  - ransomware_protection: 1.4.0
  - recommendations: 0.4.0
  - registration: 0.4.7
  - richdocuments: 3.3.15
  - serverinfo: 1.6.0
  - sharebymail: 1.6.0
  - sharerenamer: 2.7.0
  - social: 0.2.101
  - socialsharing_diaspora: 1.0.4
  - socialsharing_email: 1.0.5
  - socialsharing_facebook: 1.0.4
  - socialsharing_googleplus: 1.0.4
  - socialsharing_twitter: 1.0.4
  - spreed: 6.0.4
  - support: 1.0.0
  - survey_client: 1.4.0
  - suspicious_login: 1.0.0
  - systemtags: 1.6.0
  - tasks: 0.11.1
  - telephoneprovider: 1.0.2
  - terms_of_service: 1.2.2
  - text: 1.0.2
  - theming: 1.7.0
  - theming_customcss: 1.3.0
  - twofactor_backupcodes: 1.5.0
  - updatenotification: 1.6.0
  - video_converter: 0.1.0
  - viewer: 1.1.0
  - weather: 1.6.1
  - workflowengine: 1.6.0
Disabled:
  - encryption
  - files_antivirus
  - files_external
  - files_texteditor
  - user_ldap
```

**Nextcloud configuration:**

```
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "cloud.ruppssites.com"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "16.0.4.1",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "smtp",
        "mail_smtpauthtype": "LOGIN",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "587",
        "mail_smtpauth": 1,
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "maintenance": false,
        "memcache.local": "\\OC\\Memcache\\APCu",
        "theme": "",
        "loglevel": 0,
        "default_language": "en",
        "enable_previews": true,
        "preview_max_x": 2048,
        "preview_max_y": 2048,
        "preview_max_scale_factor": 10,
        "preview_max_filesize_image": 50,
        "preview_libreoffice_path": "\/usr\/bin\/libreoffice",
        "preview_office_cl_parameters": " --headless --nologo --nofirststartwizard --invisible --norestore --convert-to pdf --outdir ",
        "enabledPreviewProviders": [
            "OC\\Preview\\PNG",
            "OC\\Preview\\JPEG",
            "OC\\Preview\\GIF",
            "OC\\Preview\\BMP",
            "OC\\Preview\\XBitmap",
            "OC\\Preview\\SVG",
            "OC\\Preview\\TIFF",
            "OC\\Preview\\MP3",
            "OC\\Preview\\TXT",
            "OC\\Preview\\Font",
            "OC\\Preview\\MarkDown",
            "OC\\Preview\\Movie",
            "OC\\Preview\\MSOffice2003",
            "OC\\Preview\\MSOffice2007",
            "OC\\Preview\\MSOfficeDoc",
            "OC\\Preview\\OpenDocument",
            "OC\\Preview\\PDF",
            "OC\\Preview\\Postscript",
            "OC\\Preview\\StarOffice"
        ],
        "tempdirectory": "\/home\/support\/domains\/cloud.ruppssites.com\/tmp\/nextcloudtemp",
        "overwrite.cli.url": "https:\/\/cloud.ruppssites.com\/",
        "updater.release.channel": "production",
        "has_rebuilt_cache": true
    }
}
```

**Are you using external storage, if yes which one:** no.

**Are you using encryption:** yes https, but not end to end.

**Are you using an external user-backend, if yes which one:** no.

### Client configuration
**Browser:** Chrome for Ubuntu Linux latest.

**Operating system:** Ubuntu 18.04 workstation

Logs aren't needed because it's not really working wrong by software bugs, it's working wrong because of screen management issues.

### LINKED ISSUE #787: Feature request: All bookmarks
**Is your feature request related to a problem? Please describe.**
1. When you want to view all your bookmarks, first of all, it takes ages to get through the list because of the infinite scroll, takes too long to load "pages" and  "pages", please add a button or some other way to load it all at once.
2. Folders do not show under "All bookmarks", the interface is really confusing, it looks like a flat view w/o folders.

### PR DESCRIPTION
**Pull request by marcelklehr**
fixes #543 
fixes #694
fixes #756 
fixes #633 
fixes #772
fixes #747
fixes #784
fixes #787

- [ ] Fix the content calc 100% @skjnldsv

### CODE REVIEW

**Comment by marcelklehr** [comment]
Bookmarks - Nextcloud (5)
Bookmarks - Nextcloud (4)
Bookmarks - Nextcloud (3)
Bookmarks - Nextcloud (2)
Bookmarks - Nextcloud (1)

/cc @nextcloud/designers -- Feedback appreciated :)

File: webpack.common.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+				options: {
+					name: '[name].[ext]?[hash]'
+				}
+			}
+		]
+	},
+	plugins: [
+		new VueLoaderPlugin(),
+		new StyleLintPlugin(),
+		new webpack.IgnorePlugin(/^\.\/locale$/, /moment$/)
```
Comment (skjnldsv) [inline webpack.common.js]: Are you using moment?

File: src/router.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import Vue from 'vue';
+import Router from 'vue-router';
+import ViewPrivate from './components/ViewPrivate';
+
+Vue.use(Router);
+
+export default new Router({
+	mode: 'hash',
+	base: OC.generateUrl('/apps/bookmarks'),
```
Comment (skjnldsv) [inline src/router.js]: Use https://github.com/ChristophWurst/nextcloud-router/ :)

`import { generateUrl } from 'nextcloud-router'`

File: src/main.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import AppGlobal from './mixins/AppGlobal';
+
+Vue.mixin(AppGlobal);
+Vue.directive('tooltip', Tooltip);
+
+const BookmarksApp = (global['Bookmarks'] = new Vue({
+	el: '#content',
+	store,
+	router,
+	render: h => h(App)
```
Comment (skjnldsv) [inline src/main.js]: indentation

File: src/main.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+import '@babel/polyfill/noConflict';
+
+import Vue from 'vue';
+import { Tooltip } from 'nextcloud-vue';
```
Comment (skjnldsv) [inline src/main.js]: If you want to be lighter, you can import the directive only, it should avoid importing the full library. `import Tooltip from 'nextcloud-vue/dist/Directives/Tooltip'`

File: src/main.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+import '@babel/polyfill/noConflict';
```
Comment (skjnldsv) [inline src/main.js]: You should not need babel polyfill anymore when using babel preset-env :)

File: src/components/ViewPrivate.vue [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+				window.scrollY + window.innerHeight + 500
+			) {
+				this.$store.dispatch(actions.FETCH_PAGE);
+			}
+		}
+	}
+};
+</script>
+<style>
+#app-content {
```
Comment (skjnldsv) [inline src/components/ViewPrivate.vue]: You should not need to specify width here, what was the issue you faced?

File: src/components/ViewPrivate.vue [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+	},
+	computed: {
+		bookmarks() {
+			return this.$store.state.bookmarks;
+		},
+		folders() {
+			return this.$store.state.folders;
+		},
+		tags() {
+			return this.$store.state.tags;
```
Comment (skjnldsv) [inline src/components/ViewPrivate.vue]: @nextcloud/vue what do you think about this (accessing state directly) and not using getters?

File: src/components/TreeFolder.vue [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+			required: true
+		}
+	},
+	data() {
+		return { showChildren: false };
+	}
+};
+</script>
+<style>
+.Bookmarks__TreeFolder__Title {
```
Comment (skjnldsv) [inline src/components/TreeFolder.vue]: Please no caps in classes :)
Also, if you can, we usually use BEM syntax on our vue library, contacts... It looks pretty good I think, what do you think? :wink:

File: src/components/Settings.vue [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+		},
+		async onClearData() {
+			if (
+				!confirm(
+					t('bookmarks', 'Do you really want to delete all your bookmarks?')
+				)
+			) {
+				return;
+			}
+			await this.$store.dispatch(actions.DELETE_BOOKMARKS);
```
Comment (skjnldsv) [inline src/components/Settings.vue]: I like the actions constants! :+1:

File: src/components/SidebarBookmark.vue [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+					<span class="icon-calendar-dark" />
+					{{ t('bookmarks', 'Creation date') }}
+				</h3>
+				{{ addedDate }}
+			</div>
+			<div>
+				<h3><span class="icon-tag" /> {{ t('bookmarks', 'Tags') }}</h3>
+				<Multiselect
+					class="Bookmarks__Sidebar__Tags"
+					:value="tags"
```
Comment (skjnldsv) [inline src/components/SidebarBookmark.vue]: indentation

File: src/components/SidebarBookmark.vue [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+					:autoLimit="false"
+					:limit="7"
```
Comment (skjnldsv) [inline src/components/SidebarBookmark.vue]: indentation

File: src/components/SidebarBookmark.vue [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+					:options="allTags"
+					:multiple="true"
+					:taggable="true"
+					@input="onTagsChange"
+					@tag="onAddTag"
+				>
+					<template #option="{ option }">
+						<span>{{ option | tagLabel }}</span>
```
Comment (skjnldsv) [inline src/components/SidebarBookmark.vue]: Not needed as fixed in the library already :)

File: src/components/Settings.vue [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+	},
+	methods: {
+		onImportOpen(e) {
+			e.target.previousElementSibling.click();
+		},
+		onImportSubmit(e) {
+			this.$store.dispatch(actions.IMPORT_BOOKMARKS, e.target.files[0]);
+		},
+		onExport() {
+			window.location =
```
Comment (skjnldsv) [inline src/components/Settings.vue]: Seems like a job for this.$router? :)

File: src/components/Settings.vue [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+	name: 'Settings',
+	components: {},
+	computed: {
+		oc_defaults() {
+			return window.oc_defaults;
+		},
+		bookmarklet() {
+			const bookmarkletUrl =
+				window.location.origin + OC.generateUrl('/apps/bookmarks/bookmarklet');
+			return `javascript:(function(){var a=window,b=document,c=encodeURIComponent,e=c(document.title),d=a.open('${bookmarkletUrl}?output=popup&url='+c(b.location)+'&title='+e,'bkmk_popup','left='+((a.screenX||a.screenLeft)+10)+',top='+((a.screenY||a.screenTop)+10)+',height=500px,width=550px,resizable=1,alwaysRaised=1');a.setTimeout(function(){d.focus()},300);})();`;
```
Comment (skjnldsv) [inline src/components/Settings.vue]: This seems like a bad thing to do :thinking: 
Please use @click :)

File: src/components/Navigation.vue [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+				...this.tagMenu
+			];
+		}
+	},
+
+	created() {},
+
+	methods: {
+		onNewBookmark() {
+			this.$store.commit(
```
Comment (skjnldsv) [inline src/components/Navigation.vue]: indentation

File: src/components/Breadcrumbs.vue [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+			></button>
+		</div>
+	</div>
+</template>
+<script>
+import { Multiselect } from 'nextcloud-vue';
+import { actions, mutations } from '../store';
+
+export default {
+	name: 'Breadcrumbs',
```
Comment (skjnldsv) [inline src/components/Breadcrumbs.vue]: This deserves to have its component in the vue library :)
Could you create it there? :rocket:

File: src/components/BookmarksListFolder.vue [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+	props: {
+		folder: {
+			type: Object,
+			required: true
+		}
+	},
+	data() {
+		return { renaming: false, title: this.folder.title };
+	},
+	created() {},
```
Comment (skjnldsv) [inline src/components/BookmarksListFolder.vue]: Unneeded?

File: package.json [hunk 18]
```diff
@@ -13,16 +19,48 @@
...
-		"interactjs": "^1.5.3",
-		"jquery": "^3.3.1",
-		"select2": "^4.0.6",
-		"underscore": "^1.9.1"
+		"@babel/polyfill": "^7.4.4",
+		"humanize-duration": "^3.20.1",
+		"nextcloud-axios": "^0.2.0",
+		"nextcloud-server": "^0.15.10",
+		"nextcloud-vue": "^0.12.0",
+		"uuid": "^3.3.2",
```
Comment (skjnldsv) [inline package.json]: Unused?

File: .eslintrc.js [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+		parser: 'babel-eslint',
+		ecmaVersion: 6
+	},
+	extends: [
+		'eslint:recommended',
+		'plugin:import/errors',
+		'plugin:import/warnings',
+		'plugin:node/recommended',
+		'plugin:vue/essential',
+		//'plugin:vue/recommended',
```
Comment (skjnldsv) [inline .eslintrc.js]: Needed :)
Also, please enable `plugin:nextcloud/recommended` https://github.com/skjnldsv/vueexample/blob/b16df62098ecd647b1f4300c1e5b889f92d48ef1/.eslintrc.js#L28

File: src/components/Settings.vue (hunk 14 above)
Comment (ChristophWurst) [reply src/components/Settings.vue]: > Please use @click :)

nice @mention

**Review by skjnldsv — CHANGES_REQUESTED** [review]
Great work!!
I did a first pass, don't hesitate to ask if I did a mistake! :wink: 

I see a lot of indentations issues, Did you do a lint fix before pushing?

Also, the icons in the breadcrumbs are too small, are you using the Actions component? :)

File: src/components/Settings.vue (hunk 14 above)
Comment (marcelklehr) [reply src/components/Settings.vue]: Ah, this is not supposed to be run in the current window context. It's a bookmarklet URL. You drag the bookmarklet into your browser toolbar ;)

**Comment by marcelklehr** [comment]
I'm aware of the indentation/styling issues. This is mostly because I'm using prettier, which seems to conflict with lots of styling related eslint rules :/ It Would be really great to have both, but I couldn't bring myself to sort this out just yet :)

File: src/components/ViewPrivate.vue (hunk 6 above)
Comment (marcelklehr) [reply src/components/ViewPrivate.vue]: image

vs 

image

File: src/components/TreeFolder.vue (hunk 8 above)
Comment (marcelklehr) [reply src/components/TreeFolder.vue]: Sure, I'm not really a css guru :)

**Comment by marcelklehr** [comment]
> Also, the icons in the breadcrumbs are too small, are you using the Actions component? :)

Ah, nope, I am not. Good idea :)

File: src/components/Settings.vue (hunk 14 above)
Comment (skjnldsv) [reply src/components/Settings.vue]: Ah, sorry! ):

File: src/components/Settings.vue (hunk 13 above)
Comment (marcelklehr) [reply src/components/Settings.vue]: This is used to trigger a download, though. I'm not sure if you can do that with $router.

File: src/components/ViewPrivate.vue (hunk 6 above)
Comment (skjnldsv) [reply src/components/ViewPrivate.vue]: Let's leave it like that I'll have a look later :)
Let's create an issue when we'll merge this

File: src/components/SidebarBookmark.vue (hunk 12 above)
Comment (marcelklehr) [reply src/components/SidebarBookmark.vue]: :tada:

**Comment by jancborchardt** [comment]
O lala @marcelklehr this looks super cool, great work! :) Really only have some detail feedback:
- [ ] The grid/list view switcher could be just 1 icon, like in Files: showing the view that you will switch to, no button background or border (only on hover) and slightly transparent.
- [ ] The grid view elements could have `border-radius: var(--border-radius);`
- [ ] In grid view, the placement of the tags in the top left looks a bit odd. Not sure, I would probably put them below the title of the element. But maybe it’s ok like it is cause putting them below the title just creates a whole bunch of other questions like how does it look like without them etc. Maybe directly above the title would be good?
- [ ] Get a drink and chill cause that’s some fine work :D

**Comment by marcelklehr** [comment]
@skjnldsv Thanks for all the helpful input!
@jancborchardt Oooh, the border-radius is beautiful!

I agree somewhat on the tags. They don't seem to fit too well anywhere else, either, though. Here are some options:


Screenshot from 2019-08-26 19-57-53
Screenshot from 2019-08-26 19-57-21
Screenshot from 2019-08-26 20-01-00

**Comment by jancborchardt** [comment]
> 

This one is definitely the best, all info in one spot. :) Maybe even make them `opacity:.5`, but might not even be necessary.

File: src/components/Settings.vue [hunk 20]
```diff
@@ -0,0 +1,177 @@
...
+        {{ t('bookmarks', 'Delete all bookmarks') }}
+      </button>
+    </label>
+
+    <label>{{ t('bookmarks', 'Bookmarklet') }}
+      <a v-tooltip="t('bookmarks', 'Drag this to your browser bookmarks and click it to quickly bookmark a webpage')" class="button" :href="bookmarklet"
+         @click.prevent="void 0"
+      >{{
+        t('bookmarks', 'Add to {instanceName}', {
+          instanceName: oc_defaults.name
```
Comment (skjnldsv) [inline src/components/Settings.vue]: @ChristophWurst do we have something for that on our libs?

**Comment by marcelklehr** [comment]
Mmmh. @skjnldsv I still have the following result with nextcloud-vue@0.12.1:

Screenshot from 2019-08-29 13-56-39

**Comment by skjnldsv** [comment]
@marcelklehr update to 0.12.2

File: src/components/Settings.vue (hunk 20 above)
Comment (marcelklehr) [reply src/components/Settings.vue]: I'll take that as a no?

**Comment by marcelklehr** [comment]
@skjnldsv Would you be amendable to checking out the UX of this new version as well? I'd love to hear your feedback on things like creating and moving bookmarks/folders.

**Comment by skjnldsv** [comment]
@marcelklehr sure!!
I'll do another pass on the reviews this week :)

**Comment by skjnldsv** [comment]
Awesome job so far! :)

A few remarks/issues:
- Navigation:
  - Nothing so far: Looks great!
- Settings:
  - When switching to list view, and changing folder, the select is back to grid view despite the ui being list view :)
- Main content:
  - Default tags multiselect should have "select option(s)" as placeholder as we cans select multiple
  - Default list view entry looks really empty, could we add the description inline?
  Capture d’écran_2019-09-09_10-51-51
  - 
- Sidebar:
  - If we're not displaying a preview of the page, and just the icon, we should use the compact view of the sidebar. https://nextcloud-vue-components.netlify.com/#/Components?id=appsidebar
  Capture d’écran_2019-09-09_10-47-05
  - Ideally we should display the preview (rich previews like facebook, twitter... etc, is it hard? Maybe for another pr ;) )
  - Sharing tab is empty 
  Capture d’écran_2019-09-09_10-48-08
  - If no tags, text in multiselect should give the hint the user can type to add a new tag. https://github.com/nextcloud/server/blob/f27f965ad9e9621c8db6714a72419da046a1b5c4/apps/files_sharing/src/components/SharingInput.vue#L31
   Capture d’écran_2019-09-09_10-47-30

**Comment by marcelklehr** [comment]
> When switching to list view, and changing folder, the select is back to grid view despite the ui being list view :)

Ah, yes, the view mode setting is not quite intuitive, yet. So, the button in the upper right corner allows changing the vie temporarily, while the setting allows changing the default for this, ie. for when you load the page.

**Comment by marcelklehr** [comment]
> Ideally we should display the preview (rich previews like facebook, twitter... etc, is it hard? Maybe for another pr ;) )

This is already implemented, but it relies on a third-party service, so requires admin set-up

**Comment by marcelklehr** [comment]
> Default list view entry looks really empty, could we add the description inline?

nice idea! :+1:

**Comment by marcelklehr** [comment]
> Sharing tab is empty

Yup, this is the next feature :grinning:

**Comment by jancborchardt** [comment]
> Ah, yes, the view mode setting is not quite intuitive, yet. So, the button in the upper right corner allows changing the vie temporarily, while the setting allows changing the default for this, ie. for when you load the page.

Just like we do it in the Files app, there don’t need to be 2 separate things here, as you see this causes confusion. ;) We only need the toggle on the top right → it at the same time allows you to switch it quickly, and it should be permanent. Makes sense? :)

**Comment by marcelklehr** [comment]
Thanks for all you feedback :relaxed:

**Comment by skjnldsv** [comment]
Closed by mistake?

**Comment by marcelklehr** [comment]
Nope, I merged on the terminal and had to rebase, which presumably confused github.

**Comment by jancborchardt** [comment]
Heh, was already worried. :D Good stuff @marcelklehr! :cake:

**Pull request closed without merging**